src/helics/core/mpi/MpiService.cpp

```diff
@@ -0,0 +1,314 @@
+All rights reserved.
+
+This software was co-developed by Pacific Northwest National Laboratory, operated by the Battelle Memorial Institute; the National Renewable Energy Laboratory, operated by the Alliance for Sustainable Energy, LLC; and the Lawrence Livermore National Laboratory, operated by Lawrence Livermore National Security, LLC.
+
+*/
+
+#include "MpiService.h"
+
+namespace helics {
+
+MPI_Comm MpiService::mpiCommunicator = MPI_COMM_NULL;
+
+MpiService& MpiService::getInstance ()
+{
+    static MpiService instance;
+    instance.startService ();
+    return instance;
+}
+
+void MpiService::setMpiCommunicator (MPI_Comm communicator)
+{
+    mpiCommunicator = communicator;
+}
+
+MpiService::MpiService ()
+    : commRank (-1)
+{
+    comms_connected = 0;
+    stop_service = false;
+    startup_flag = false;
+}
+
+MpiService::~MpiService ()
+{
+    // Stop the service thread
+    stop_service = true;
+    if (service_thread->joinable ())
+    {
+        service_thread->join ();
+    }
+}
+
+void MpiService::startService ()
+{
+    if (service_thread == nullptr)
+    {
+        startup_flag = true;
+        service_thread = std::make_unique<std::thread> (&MpiService::serviceLoop, this);
+        while (startup_flag && !stop_service);
+    }
+}
+
+void MpiService::serviceLoop ()
+{
+    // Startup/teardown for MPI can all go in separate functions
+    // For integrating helics MPI calls with existing MPI applications, user may want to call from their own thread
+
+    // Initialize MPI
+    if (initMPI ())
+    {
+        // if mpiCommunicator isn't set by user, make it a duplicate of MPI_COMM_WORLD
+        if (mpiCommunicator == MPI_COMM_NULL)
+        {
+            MPI_Comm_dup (MPI_COMM_WORLD, &mpiCommunicator);
+        }
+
+        // set commRank to our process rank
+        MPI_Comm_rank (mpiCommunicator, &commRank);
+    }
+
+    // signal that we have finished starting
+    startup_flag = false;
+
+    std::cout << "Started MPI service loop for rank " << commRank << std::endl;
+
+
+    // Run as long as we have something in the send queue or the chance of getting something
+    while (!stop_service || comms_connected > 0 || !(txMessageQueue.empty ()))
+    {
+        // send/receive MPI messages
+        sendAndReceiveMessages ();
+    }
+
+    // Post receives for any waiting sends
+    int message_waiting = 1;
+    MPI_Status status;
+    while (message_waiting)
+    {
+        MPI_Iprobe (MPI_ANY_SOURCE, MPI_ANY_TAG, mpiCommunicator, &message_waiting, &status);
+        if (message_waiting)
+        {
+            // Get the size of the message waiting to be received
+            int recv_size;
+            std::vector<char> buffer;
+            MPI_Get_count (&status, MPI_CHAR, &recv_size);
+            buffer.resize (recv_size);
+            
+            // Receive the message
+            MPI_Recv (buffer.data (), buffer.capacity (), MPI_CHAR, MPI_ANY_SOURCE, MPI_ANY_TAG, mpiCommunicator, &status);
+        }
```

Comment:
I understand you're trying to drain and discard any pending messages from MPI.  If I were being pedantic, I would say to use the status.MPI_SOURCE and status.MPI_TAG values from the MPI_Status object in the MPI_Recv(), instead of again specifying MPI_ANY_SOURCE and MPI_ANY_TAG.  There is a slight possibility that MPI could reorder the messages and you'd be receiving a different message than what probe indicated, so your size/capacity could be wrong.